api/test/nostd/list_test.cc

```diff
@@ -0,0 +1,132 @@
+#include "opentelemetry/nostd/list.h"
+#include "opentelemetry/nostd/pair.h"
+
+#include <gtest/gtest.h>
+
+using opentelemetry::nostd::list;
+using opentelemetry::nostd::pair;
+
+/* Tests that front() returns the first element and back()
+ * returns the last element.
+ */
+TEST(ListTest, FrontBack)
+{
+  list<int> list_1;
+
+  list_1.push_back(9);
+  list_1.push_back(8);
+  list_1.push_back(7);
+  list_1.push_back(6);
+
+  EXPECT_EQ(list_1.front(), 9);
+  EXPECT_EQ(list_1.back(), 6);
+}
+
+/* Tests that the list object can handle data types of a size
+ * other than an int. */
+TEST(ListTest, PairList)
+{
+  list<pair<int, int>> pair_list;
+
+  pair_list.push_back(pair<int, int>(1, 2));
+  pair_list.push_back(pair<int, int>(2, 3));
+  pair_list.push_back(pair<int, int>(4, 5));
+
+  pair<int, int> temp_pair;
+  temp_pair = pair_list.front();
+
+  EXPECT_EQ(pair_list.front().first(), 1);
+  EXPECT_EQ(pair_list.front().second(), 2);
+
+  EXPECT_EQ(pair_list.back().first(), 4);
+  EXPECT_EQ(pair_list.back().second(), 5);
+}
+
+/* Tests taht a list object can be initialized with an initalizer list. */
+TEST(ListTest, InitializerList)
+{
+
+  list<int> pair_list = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+
+  EXPECT_EQ(pair_list.front(), 1);
+  EXPECT_EQ(pair_list.back(), 10);
+}
+
+/* Tests that the list object can handle an insert_iterator. */
+TEST(ListTest, IteratorInsertion)
+{
+
+  list<int> int_list    = {1, 2, 3, 4};
+  list<int> insert_list = {5, 6, 7, 8};
+
+  std::insert_iterator<list<int>> back(int_list, std::end(int_list));
+
+  for (auto iter = std::begin(insert_list); iter != std::end(insert_list); ++iter)
+  {
+    back = *iter;
+  }
+
+  int counter = 1;
+  for (auto iter = std::begin(int_list); iter != std::end(int_list); ++iter)
+  {
+    EXPECT_EQ(*iter, counter);
```

Comment:
I am having some trouble finding the header that contains ElementsAre

Reply:
It should be in gmock/gmock.h

Reply:
FYI: https://github.com/google/googletest/blob/master/googlemock/docs/cheat_sheet.md#container-matchers

Reply:
I have tried gmock/gmock.h and am getting "'ElementsAre' was not declared in this scope"

Reply:
Probably you're missing the namespace? testing::ElementsAre

Reply:
It looks like in order to use ElementsAre, I would also need to implement a const iterator for nostd::list. Do you think that would be valuable?

Reply:
I wouldn't really recommend extending the nostd::list interface specifically for testing, but a const iterator seems like it would be useful generally. Do it if you think it will be quick and you have some spare time.